src/cswinrt/code_writers.h

```diff
@@ -6340,8 +6419,117 @@ type.TypeNamespace(),
+                    if (std::holds_alternative<generic_type_param>(semantics))
+                    {
+                        genericArgs.push_back(w.write_temp("%", bind<write_generic_type_name>(i)));
+                    }
+                    i++;
+                }                
+                if (genericArgs.size() == 0)
+                {
+                    return;
+                }
+                w.write("<%>", bind_list([](writer& w, auto&& value)
+                    {
+                        w.write(value);
+                    }, ", "sv, genericArgs));
+            });
+    }
+
+    void write_event_source_subclass(writer& w, cswinrt::type_semantics eventTypeSemantics)
+    {
+        for_typedef(w, eventTypeSemantics, [&](TypeDef const& eventType)
+            {
+                if (eventType.TypeNamespace() == "System" && eventType.TypeName() == "EventHandler`1")
+                {
+                    return;
+                }
+                auto eventTypeCode = w.write_temp("%", bind<write_type_name>(eventType, typedef_name_type::Projected, false));
+                auto invokeMethodSig = get_event_invoke_method(eventType);
+                w.write(R"(
+    internal unsafe class %% : EventSource<%>
+    {
+        private % handler;
+
+        internal %(IObjectReference obj,
+            delegate* unmanaged[Stdcall]<System.IntPtr, System.IntPtr, out WinRT.EventRegistrationToken, int> addHandler,
+            delegate* unmanaged[Stdcall]<System.IntPtr, WinRT.EventRegistrationToken, int> removeHandler) : base(obj, addHandler, removeHandler)
+        {
+        }
+
+        override protected System.Delegate EventInvoke
+        {
+            get
+            {
+                if (handler == null)
+                {
+                    handler = (%) =>
+                    {
+                        if (_event == null)
+                        {
+                            return %;
+                        }
+                        %_event.Invoke(%);
+                    };
+                }
+                return handler;
+            }
+        }
+    }
+)",
+                    bind<write_event_source_type_name>(eventTypeSemantics),
+                    bind<write_event_source_generic_args>(eventTypeSemantics),
+                    eventTypeCode, 
+                    eventTypeCode,
+                    bind<write_event_source_type_name>(eventTypeSemantics), 
+                    bind<write_event_invoke_params>(invokeMethodSig),
+                    bind<write_event_invoke_return_default>(invokeMethodSig),
+                    bind<write_event_invoke_return>(invokeMethodSig),
+                    bind<write_event_invoke_args>(invokeMethodSig));
+                });
+    }
+
+    void write_temp_class_event_source_subclass(writer& w, TypeDef const& classType, concurrency::concurrent_unordered_map<std::string, std::string>& typeNameToDefinitionMap)
+    {
+        for (auto&& ii : classType.InterfaceImpl())
```

Comment:
Can you remind me in which scenario this would add a event source class which isn't already added by write_temp_interface_event_source_subclass based on the interface declaration and is used.  The only scenario I see here is a class is declared which implements an interface from another projection.  But in that case would the event source be instantiated in this projection or would it be in the other projection in the interface ABI helper type?  I might be missing a scenario where this is used.

Reply:
This is helpful when the class has fixed generic types but interface has variable. The following scenario for example:
```
interface Vector<T>
{
    ...
}

class MyVector<string> implements Vector<string>
{
   ...
}
```
In this case Vector<T> and Vector<string> can have different events